File: wllib/src/main/java/com/wl/wllib/LogToFile.java
package com.wl.wllib;

import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Message;
import android.text.TextUtils;
import android.util.Log;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;

/**
 * 日志模块，
 * 会将日志缓存到根目录下的MyLog1.txt和MyLog2.txt和
 * 每个文件默认最大为200M
 */
public class LogToFile extends HandlerThread {
    private static Handler mHandler;
    private static LogToFile mLogToFile;
    private static final String LOG_THREAD = "LOG_THREAD";
    private static final String TAG = "LogToFile";

    private static boolean isDebug = true;

    public boolean isDebug() {
        return isDebug;
    }

    public void setDebug(boolean debug) {
        isDebug = debug;
    }

    private LogToFile(String name) {
        super(name);
    }


    //在application初始化 LogToFile.init();
    public static void init(boolean debug) {
        isDebug = debug;
        if (mLogToFile == null) {
            mLogToFile = new LogToFile(LOG_THREAD);
            mLogToFile.start();
            mHandler = new Handler(mLogToFile.getLooper()) {
                @Override
                public void handleMessage(Message msg) {
                    super.handleMessage(msg);
                    if (msg != null) {//子线程执行保存文件操作
                        Bundle data = msg.getData();
                        saveLogFile(data.getString("TAG", "tag is null"), data.getString("MESSAGE", "message is null"));
                    }
                }
            };
        }
    }

    //在程序退出时调用
    public static void deInit() {
        if (mLogToFile != null) {
            mHandler.removeCallbacksAndMessages(null);
            mLogToFile.quit();
            mLogToFile = null;
            mHandler = null;
        }
    }

    public static void v(String tag, String msg) {
        if (isDebug) {
            Log.d(tag, msg);
        }
        toChildThread(tag, msg);
    }

    public static void d(String tag, String msg) {
        if (isDebug) {
            Log.d(tag, msg);
        }
        toChildThread(tag, msg);
    }

    public static void d(String msg) {
        if (isDebug) {
            Log.d(TAG, msg);
        }
        toChildThread(TAG, msg);
    }

    public static void i(String tag, String msg) {
//        if (isDebug) {
//            Log.i(tag, msg);
//        }
        toChildThread(tag, msg);
    }

    public static void i(String msg) {
        if (isDebug) {
            Log.i(TAG, msg);
        }
        toChildThread(TAG, msg);
    }


    public static void w(String tag, String msg) {
        if (isDebug) {
            Log.w(tag, msg);
        }
        toChildThread(tag, msg);
    }

    public static void e(String tag, String msg) {
        if (isDebug) {
            Log.e(tag, msg);
        }
        toChildThread(tag, msg);
    }

    public static void e(String msg) {
        if (isDebug) {
            Log.e(TAG, msg);
        }
        toChildThread(TAG, msg);
    }

    private static void toChildThread(String tag, String message) {

        if (!isDebug) {//release版本不保存
            return;
        }

        if (mHandler != null && !TextUtils.isEmpty(message)) {
            Message obtain = Message.obtain();
            Bundle bundle = new Bundle();
            if (TextUtils.isEmpty(tag)) {
                tag = TAG;
            }
            bundle.putString("TAG", tag);
            bundle.putString("MESSAGE", message);
            obtain.setData(bundle);
            mHandler.sendMessage(obtain);
        }
    }

    private static final int LOG_FILE_SIZE_MAX = 200 * 1024 * 1024;
    private static final String PATH = "/sdcard/";
    private static final String NAME1 = "MyLog.txt";
    private static final String NAME2 = "MyLog2.txt";
    private static SimpleDateFormat DateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss SSS");

    static StringBuilder sb = new StringBuilder();

    private static void saveLogFile(String tag, String message) {//

        if (TextUtils.isEmpty(tag)) {
            tag = "--Tag null--";
        }

        FileOutputStream fileOutputStream = null;
        try {
            File file = new File(PATH, getLogFileName());
            if (file == null || !file.exists()) {
                file.getParentFile().mkdirs();
                file.createNewFile();
            }
            fileOutputStream = new FileOutputStream(file, true);

//            long size = fileOutputStream.getChannel().size();
//            if (size > LOG_FILE_SIZE_MAX) {//超过20m，删除
//                fileOutputStream.close();
//                file.delete();
//                return;
//            }
            sb.delete(0, sb.length());
            sb.append(DateFormat.format(new Date()));
            sb.append("/");
            sb.append(tag);
            sb.append("=====>>[");
            sb.append(message);
            sb.append("]\n");
            String result = sb.toString();
            fileOutputStream.write(result.getBytes());
            fileOutputStream.flush();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (fileOutputStream != null) {
                    fileOutputStream.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * 建立两个日志文件 Log1 Log2 轮流存储日志，当当前的文件存储到一半并且另一个文件满了时候，删除另一个
     *
     * @return
     */
    private static String getLogFileName() throws IOException {
        File file1 = new File(PATH, NAME1);
        File file2 = new File(PATH, NAME2);
        FileOutputStream fileOutputStream1 = null;
        FileOutputStream fileOutputStream2 = null;
        try {
            if (file1 == null || !file1.exists()) {
                file1.getParentFile().mkdirs();
                file1.createNewFile();
            }

            fileOutputStream1 = new FileOutputStream(file1, true);
            long size1 = fileOutputStream1.getChannel().size();

            if (file2 == null || !file2.exists()) {
                file2.getParentFile().mkdirs();
                file2.createNewFile();
            }
            fileOutputStream2 = new FileOutputStream(file2, true);
            long size2 = fileOutputStream2.getChannel().size();


            if (size1 == 0) {//file1 0
                if (size2 == 0 || size2 >= LOG_FILE_SIZE_MAX) {
                    return NAME1;
                } else {
                    return NAME2;
                }
            } else if (size1 > 0) {
                if (size1 >= LOG_FILE_SIZE_MAX) {
                    if ((size2 >= LOG_FILE_SIZE_MAX / 1.2)) {
                        file1.delete();
                        LogToFile.e(TAG, "file1.delete() size1=" + size1 + " size2=" + size2);
                    }
                    return NAME2;
                } else {
                    if ((size1 >= LOG_FILE_SIZE_MAX / 1.2) && (size2 >= LOG_FILE_SIZE_MAX)) {
                        file2.delete();
                        LogToFile.e(TAG, "file2.delete() size1=" + size1 + " size2=" + size2);
                    }
                    return NAME1;
                }
            }
//            if (size > LOG_FILE_SIZE_MAX) {//超过20m，删除
//                fileOutputStream1.close();
//                file1.delete();
//            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            LogToFile.e(TAG, "_______________________________________________________FileNotFoundException_______________________________________________________");
        } catch (IOException e) {
            e.printStackTrace();
            LogToFile.e(TAG, "_______________________________________________________IOException_______________________________________________________");
        } finally {
            if (fileOutputStream1 != null) {
                fileOutputStream1.close();
            }
            if (fileOutputStream2 != null) {
                fileOutputStream2.close();
            }
        }
        return NAME1;
    }

    public static void test() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                while (true) {

                    LogToFile.d(TAG, "_asdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdfjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdfjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdfjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdfjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdfjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdfjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdfjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdfjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdfjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdljasdlasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdljasdlasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdljasdlasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlljasdlasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlljasdlasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlljasdlasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdlasdfkl;ajsdfl;jasdl;fjkasdlkvmck,.zxjcklisuafodjasklfdjasldfjasdl" +
                            "_______________________________________________________");
                    try {
                        Thread.sleep(200);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        }).start();
    }
}

File: wllib/src/main/java/com/wl/wllib/LogToFile.kt
package com.wl.wllib

import android.util.Log
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.max
import kotlin.math.min

object LogToFile {
    @JvmStatic
    var fileName1 = "/sdcard/MyLog10.txt"

    @JvmStatic
    var fileName2 = "/sdcard/MyLog11.txt"
    private var file1: File? = null
    private var file2: File? = null
    private var curFile: File? = null
    var MaxSize = 50 * 1024 * 1024

    val DateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss SSS")
    private var fos: FileOutputStream? = null

    @JvmStatic
    var format: Format? = null

    enum class Code {
        User,//用户操作
        Cmd, //与下位机交互
        Info,//运行日志
        Error//异常报错
    }

    private var isAllowInput = true
    private var tsHashMap = hashMapOf<String, TSModel>()

    @JvmStatic
    fun stopInput() {
        isAllowInput = false
        fos?.let {
            synchronized(it) {
                fos?.close()
            }
        }
    }

    @JvmStatic
    fun startInput() {
        init()
        isAllowInput = true
        putMsgTs()
    }

    private fun putMsgTs() {
        for ((k, v) in tsHashMap) {
            input(v.code, v.tag, v.msg)
        }
        tsHashMap.clear()
    }

    @JvmStatic
    fun init() {
        file1 = File(fileName1)
        file2 = File(fileName2)
        if (!file1!!.exists()) {
            file1!!.parentFile.mkdirs()
            file1!!.createNewFile()
        }
        if (!file2!!.exists()) {
            file2!!.parentFile.mkdirs()
            file2!!.createNewFile()
        }
        curFile = getFile()
        fos = FileOutputStream(curFile, true)
    }

    /**
     * 获取这次输出的文件
     * 如果文件大于最大值并且另一个文件大于1/3最大值，就删除这个文件
     * 1、如果两个文件有大于最大值的直接删除
     * 2、返回输出的文件
     * 2.1、file1Size == 0 && file2Size 不为空 return file2
     * 2.2、file2Size == 0 && file1Size 不为空 return file1
     * 2.3、两个都不为空 return 最小的文件
     * @return File
     */
    fun getFile(): File {
        var file1Size = file1!!.length()
        var file2Size = file2!!.length()

        if (file1Size > MaxSize && file2Size > MaxSize / 3) {
            file1!!.delete()
            file1!!.createNewFile()
            file1Size = 0
        } else if (file2Size > MaxSize && file1Size > MaxSize / 3) {
            file2!!.delete()
            file2!!.createNewFile()
            file2Size = 0
        }
        return if (file1Size == 0L && file2Size in 1 until MaxSize) {
            file2!!
        } else if (file2Size == 0L && file1Size in 1 until MaxSize) {
            file1!!
        } else {
            min(file1Size, file2Size).let {
                if (it == file2Size) {
                    return file2!!
                } else {
                    return file1!!
                }
            }
        }
    }

    @JvmStatic
    fun input(code: String, tag: String, msg: String) {
        format(code, tag, msg).let {
            Log.d(tag, "$code>>>[$msg]")
            if (isAllowInput) {
                write("$it\n")
            } else {
                tsHashMap.put(code, TSModel(code, tag, msg))
            }
        }
    }

    @JvmStatic
    fun u(msg: String) {
        u(getTag(), msg)
    }

    @JvmStatic
    fun u(tag: String, msg: String) {
        input(Code.User.name, tag, msg)
    }

    @JvmStatic
    fun i(msg: String) {
        i(getTag(), msg)
    }

    @JvmStatic
    fun i(tag: String, msg: String) {
        input(Code.Info.name, tag, msg)
    }

    @JvmStatic
    fun c(msg: String) {
        c(getTag(), msg)
    }

    @JvmStatic
    fun c(tag: String, msg: String) {
        input(Code.Cmd.name, tag, msg)
    }

    @JvmStatic
    fun e(msg: String) {
        e(getTag(), msg)
    }

    @JvmStatic
    fun e(tag: String, msg: String) {
        input(Code.Error.name, tag, msg)
    }

    private fun format(code: String, tag: String, msg: String): String {
        if (format == null) format = defaultFormat
        return format!!.onFormat(code, tag, msg)
//        return "${DateFormat.format(Date())}>${tag}/${code}>>>[${msg}]"
    }

    @JvmStatic
    fun write(msg: String) {
        fos?.let {
            synchronized(it) {
                it.write(msg.toByteArray())
                //强制内存缓冲区内的数据写入到系统。不保证写入到磁盘
                it.flush()
                //强制写入此流的数据到磁盘
                it.fd.sync()
            }
        }
    }

    private fun getTag(): String {
        val sts = Throwable().stackTrace
        val jk = sts.find {
            it.className.split(".").last() != this.javaClass.simpleName
        }?.className?.split(".")?.last() ?: this.javaClass.simpleName
        return jk
    }

    interface Format {
        fun onFormat(code: String, tag: String, msg: String): String
    }

    private val defaultFormat = object : Format {
        override fun onFormat(code: String, tag: String, msg: String): String {
            return "${DateFormat.format(Date())}>${tag}/${code}>>>[${msg}]"
        }
    }

    data class TSModel(val code: String, val tag: String, val msg: String)
}
